edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java

```diff
@@ -2815,6 +2815,17 @@ public void setGradebookPermissionService(GradebookPermissionService gradebookPe
 		this.gradebookPermissionService = gradebookPermissionService;
 	}
 
-	
+	@Override
+	public Object getGradebookWithGradeMappings(final Long id) {
```

Comment:
Finally I've changed the logic and avoid some of the iterators. 

Now, first I return the sitesIds and not the sites, surely saving some memory.
Later, I don't create a gradebook list, I directly work on the gradebook in the siteid's iterator if the site has gradebook.
if it is a new scale, passing scale Uuid and gradebook Uid I update all.
if it is a update, then I need to pass the id to recuperate the gradeMappings with this method: getGradebookGradeMappings(final Long gradebookId)  I need to pass the id and not the Uid to make it work. Compare gradings and update if needed and based on the booleans.

I think now is more efficient and fast. 

As now is working fine, I'm going to update the pull request with this changes, and then start to work in the API thing.